### PR TITLE
Queue now mounts on session start

### PR DESCRIPTION
**Pull request by jamyu2006**
The changes that matter:

adding a getAccessToken api route => Post request to the database

useEffect in page.tsx => makes a fetch to the getAccessToken route on mount(depends on nothing), when the access token is recieved, updates the use state of "atoken"

useEffect in page.tsx => makes a fetch to the getQueue route on atoken change(when getAccessToken revieces the token and updates it), getQueue takes in the token and fetches the existing queue in the database, this is then updated with the queue useState

useEffect in client.tsx => on initQueue change, every song is read from initQueue and passed into the songList useState

### CODE REVIEW

**Review by foleyash — CHANGES_REQUESTED** [review]
Looks like some of the available session functionality might need a little bit of work:
- We should change the testing data to work with user input for usernames
- The isHost item should always be set to false for any new user joining a session that is not theirs. 
- The user should only be able to join sessions which they have the session ID for, so we need to add some functionality for inputting a guest code to each of these available sessions. 
- The display name for each session should be something that is not the session code as well, maybe a session name set by the host, but also just using "{host_names}'s session" should be fine. 

I would omit these parts of the PR for right now (available session parts), and then just try to have the main focus of this PR be for the Queue mounting on start up. Let me know if you have any questions about this.

Some other side notes:
- Remove the changes to pg_startup.sh since this file should not change
- Remove changes to 'src/socket/server.ts' (need to uncomment await checkQueueUpdates)
- Keep a matching version of current dependencies (need to set "next": "14.1.0" in package.json to match the current version in source code)

Other than that, the functionality for initializing the Queue on mount looks great! Should be able to approve this once the changes above are implemented :)

**Review by foleyash — COMMENTED** [review]
I am realizing now that there is a initSession route in the api folder, we could potentially call this function to grab the queue data alongside hostname and all client names within the session to initialize those on connection as well. This might be more robust than grabbing that data in separate calls. In addition it would prevent the need from bringing the access token into client side memory which would inherently be more secure. I'm sorry about the confusion earlier, but I think swapping out the api calls inside of the useEffect at 'src/app/session/[id]/page.tsx' would be the best way to approach this to also initialize data for the host's name and client names.

Basically, instead of grabbing the access token and then calling getQueue, we can simply make a fetch call to 'api/sessionDB/initSession'.

**Comment by jamyu2006** [comment]
> Looks like some of the available session functionality might need a little bit of work:
> 
> * We should change the testing data to work with user input for usernames
> * The isHost item should always be set to false for any new user joining a session that is not theirs.
> * The user should only be able to join sessions which they have the session ID for, so we need to add some functionality for inputting a guest code to each of these available sessions.
> * The display name for each session should be something that is not the session code as well, maybe a session name set by the host, but also just using "{host_names}'s session" should be fine.
> 
> I would omit these parts of the PR for right now (available session parts), and then just try to have the main focus of this PR be for the Queue mounting on start up. Let me know if you have any questions about this.
> 
> Some other side notes:
> 
> * Remove the changes to pg_startup.sh since this file should not change
> * Remove changes to 'src/socket/server.ts' (need to uncomment await checkQueueUpdates)
> * Keep a matching version of current dependencies (need to set "next": "14.1.0" in package.json to match the current version in source code)
> 
> Other than that, the functionality for initializing the Queue on mount looks great! Should be able to approve this once the changes above are implemented :)

when i initially first made the pull request it was just the one commit that was for mounting the getqueue but for some reason when i pushed the other commits up it just automatically added them to the pull request, do i have to uncommit(rebase) the latest two or make a whole new pull request

**Comment by jamyu2006** [comment]
before using initSession (theres a small problem with the queue in the database actually initialling being empty will have to look at that), i realized that i can merge the accesstoken fetch and queue fetch together: mountQueue(client side now does not have access to the accesstoken directly and only the session ID is needed)

**Review by foleyash — COMMENTED** [review]
Looks like you grabbed only the commit regarding this PR, which is great.

Some more comments just because there is a lot going on in this PR:

- Is the ReplaceQueue endpoint needed here? I couldn't find where it's being called at. If it isn't being used, we should remove this from the commit history to prevent unnecessary bloating in the codebase.
- When we shift from the getAccessToken -> getQueue approach to the initSession approach, we might want to remove the getAccessToken endpoint from the commit history because it is not being used anywhere as well.
- Try to remove unnecessary comments like the commented out console.log()s throughout 'src/app/session/[id]/client.tsx'

**Comment by foleyash** [comment]
> before using initSession (theres a small problem with the queue in the database actually initialling being empty will have to look at that), i realized that i can merge the accesstoken fetch and queue fetch together: mountQueue(client side now does not have access to the accesstoken directly and only the session ID is needed)
> 
> notable changes: added mountQueue - combines getAccessToken and getQueue page.tsx - only has one useEffect to call mountQueue

I see the issue you are talking about with initSession, I'm looking into it right now. I do still think we should work from this function, since it will grab queue, host name, and all client names for us in one go.

**Comment by foleyash** [comment]
Okay, the issue with init session should be resolved, it was regarding a separate issue with host id not being set properly. You will need to rebase your changes over the most recent changes from this branch.

**Comment by jamyu2006** [comment]
now mounts the page with init session
main concerns:
songID-initQueue does not grab the songID for the songs 
queue disparity-when fetching from database from initQueue, the queue will always mount with a disparity. The getQueue/mountQueue endpoints automatically mount with the latest, if this isn't a problem ignore.
hostname-does the hostname need to be set on each mount(both for host mount and guest mounts) I see the host useeffect so I imagine not

**Comment by foleyash** [comment]
> now mounts the page with init session main concerns: songID-initQueue does not grab the songID for the songs queue disparity-when fetching from database from initQueue, the queue will always mount with a disparity. The getQueue/mountQueue endpoints automatically mount with the latest, if this isn't a problem ignore. hostname-does the hostname need to be set on each mount(both for host mount and guest mounts) I see the host useeffect so I imagine not

- You can update/change initQueue or even the GetSessionData function if needed. If you need to grab the songID, feel free to add the 'song_id' attribute to be part of the SELECT ... attributes inside of GetSessionData in the query grabbing the queue data. The schema for our SQL relations is found under 'src/database/schema.sql'.
- "when fetching from database from initQueue, the queue will always mount with a disparity. The getQueue/mountQueue endpoints automatically mount with the latest, if this isn't a problem ignore" -- this should not be an issue since the websocket server is responsible for updating the database with the most recent version of the spotify queue on a set interval before also sending out that new queue to all clients in the session. So whatever version the new user gets of the queue from initSession, it will be the same as all other users in the session.
- Hostname will be set in the database only on the host's mount during the 'api/sessionDB/addHostName' fetch, hostname will need to be gathered on the client side ONLY by guests since the SessionHost component actually doesn't end up using hostName (since they should always know based on the fact that this component only is created when isHost == true)
- I saw your comments wondering what to do with isHostNameSet. Essentially, you should only make an api call to initSession once you know that isHostNameSet == true, so you would want to add in a check for that.

**Comment by jamyu2006** [comment]
notable changes:
initSession endpoint => now makes a queue fetch straight from spotify(in the fetch pushes all the songs to the database), if the fetch to the database is successful, does not fetch from spotify and directly mounts the database queue
mountQueue endpoint => deleted endpoint
client.tsx => added a getValue function to handle the difference between songId and song_id when retrieving from spotify and the database respectively
page.tsx => now mounts on isHostNameSet(if not, the useEffect will trigger before the hostname is set)
db.ts => added song_id to the sql, added a host_id check because of mounting problems(can remove i think?)

**Review by foleyash — CHANGES_REQUESTED** [review]
Okay, I like the layout of these changes now, but I do have a couple of questions/suggestions.

- The addition of getValue() inside of 'src/app/session/[id]/client.tsx' looks good, but I think moving this function to the 'src/utils.ts' would be more appropriate, since I believe this function could be used at any point when trying to grab data from a database object. Optional: Could also implement a setter if we down this approach for the reverse direction.
- I think we might be trying to do too much inside of the initSession route. Remember, the purpose of this endpoint is to synchronize the data a new user will see upon loading a session with the data everyone else sees inside of the session. Additionally, we have the current state of the session stored within the database, and it should be the sole responsibility of the websocket server to identify and update any changes to the spotify queue inside of the database. I don't think there's anything directly wrong with your implementation, but I do think this will increase the complexity of the code since it increases the number of database modifiers unnecessarily. Remember, even if the current state of the database is lagging behind the actual state of the spotify queue, as long as all members in the session share the same state, we should be okay.
- 'src/app/session/[id]/page.tsx' looks good to me

**Comment by foleyash** [comment]
> > Okay, I like the layout of these changes now, but I do have a couple of suggestions.
> > 
> > * The addition of getValue() inside of 'src/app/session/[id]/client.tsx' looks good, but I think moving this function to the 'src/utils.ts' would be more appropriate, since I believe this function could be used at any point when trying to grab data from a database object. Optional: Could also implement a setter if we go down this approach for the reverse direction.
> > * I think we might be trying to do too much inside of the initSession route. Remember, the purpose of this endpoint is to synchronize the data a new user will see upon loading a session with the data everyone else sees inside of the session. Additionally, we have the current state of the session stored within the database, and it should be the sole responsibility of the websocket server to identify and update any changes to the spotify queue inside of the database. I don't think there's anything directly wrong with your implementation, but I do think this will increase the complexity of the code since it increases the number of database modifiers unnecessarily. Remember, even if the current state of the database is lagging behind the actual state of the spotify queue, as long as all members in the session share the same state, we should be okay.
> > * 'src/app/session/[id]/page.tsx' looks good to me
> 
> Oh do u mean that even if the initsession doesn't grab anything, u want to leave the websocket server to deal with that? (Get rid of the Spotify fetch directly since the websocket will deal with getting the right queue anyway)
> 
> Also I have a couple implementations to the websocket server that I can show you on Thursday => adding a progress emit, and modularizing sessions

1) Yes, exactly. 2) Awesome, we can look at the websocket server together Thursday. I'm rethinking some parts of the design, I don't think it's exactly right how it is currently.

**Review by foleyash — COMMENTED** [review]
Last changes before I approve:

- Remove the changes to package-lock.json, package.json, pg_startup.sh, 'src/socket/server.ts', 'src/app/api/spotify/getQueue/route.ts', 'src/app/api/sessionDB/initSession', 'src/app/page.tsx' to ensure version of dependencies doesn't change and other files that aren't part of code changes (you can use 'git checkout <commit-hash> -- <file-path1> <file-path2> . . .' to restore specific files to their state at a given commit hash
- You will know you are good when the files changed section of this PR includes only 'src/app/api/spotify/replaceQueue' (I think this api endpoint will be useful), 'src/app/session/[id]/client.tsx and page.tsx', and 'src/database/db.ts'
